### PR TITLE
net-print/brother-mfc6490cw-bin: new ebuild

### LINKED ISSUE #72: Brother MFC-6490CW CUPS and LPR support
Hi!

The [printer-drivers overlay](https://cgit.gentoo.org/proj/printer-drivers.git/) had been [recently removed from repositories.xml](https://cgit.gentoo.org/data/api.git/commit/files/overlays/repositories.xml?id=b766404a47218209081e32ab0c90ee87175eb7fd).

I do have a Brother MFC-6490CW printer/scanner (inherited from my grandpa), and I'm happy to see that [brother-overlay provides a way to install the scanner driver](https://github.com/stefan-langenmaier/brother-overlay/tree/master/media-gfx/brother-mfc6490cw-bin).

What about [CUPS](https://cgit.gentoo.org/proj/printer-drivers.git/tree/net-print/brother-mfc6490cw-cups) and [LPR](https://cgit.gentoo.org/proj/printer-drivers.git/tree/net-print/brother-mfc6490cw-lpr) drivers...?

Is it possible/useful to add them to brother-overlay?

Also, MFC-6490CW isn't listed in the [Gentoo's Brother networked printer guide](https://wiki.gentoo.org/wiki/Brother_networked_printer).

A quick CUPS/LPR drivers fix is to add the printer-drivers overlay back via a [layman's overlay xml file](https://wiki.gentoo.org/wiki/Layman#Missing_repository.xml_file):

**_/etc/layman/overlays/printer-drivers.xml_**
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE repositories SYSTEM "/dtd/repositories.dtd">
<repositories xmlns="" version="1.0">
  <repo quality="experimental" status="unofficial">
    <name>printer-drivers</name>
    <description lang="en">User-submitted ebuilds for printer drivers</description>
    <homepage>https://cgit.gentoo.org/proj/printer-drivers.git/</homepage>
    <owner type="project">
      <email>printing@gentoo.org</email>
      <name>Gentoo printing team</name>
    </owner>
    <source type="git">https://anongit.gentoo.org/git/proj/printer-drivers.git</source>
    <source type="git">git://anongit.gentoo.org/proj/printer-drivers.git</source>
    <source type="git">git+ssh://git@git.gentoo.org/proj/printer-drivers.git</source>
    <feed>https://cgit.gentoo.org/proj/printer-drivers.git/atom/</feed>
    <!-- <feed>https://cgit.gentoo.org/proj/printer-drivers.git/rss/</feed> -->
  </repo>
</repositories>
```

Another way explained in [brother-overlay's README.md](https://github.com/stefan-langenmaier/brother-overlay/blob/master/README.md) is via a [portage's ebuild repository](https://wiki.gentoo.org/wiki//etc/portage/repos.conf#External_ebuild_repository):

**_/etc/portage/repos.conf/printer-drivers.conf_**
```conf
[printer-drivers]
location = /usr/local/portage/overlays/printer-drivers
sync-type = git
sync-uri = https://anongit.gentoo.org/git/proj/printer-drivers.git
auto-sync = yes
```

What do you think? Should the Brother MFC-6490CW's [CUPS](https://cgit.gentoo.org/proj/printer-drivers.git/tree/net-print/brother-mfc6490cw-cups) and [LPR](https://cgit.gentoo.org/proj/printer-drivers.git/tree/net-print/brother-mfc6490cw-lpr) drivers be added to brother-overlay?

### PR DESCRIPTION
**Pull request by mehw**
Fixes #72.

Add Brother MFC-6490CW CUPS and LPR support.

**Please, don't hesitate to provide suggestions.**

# Credits

[printer-drivers overlay](https://cgit.gentoo.org/proj/printer-drivers.git/), especially to the original author/maintainer (see [metadata.xml](https://cgit.gentoo.org/proj/printer-drivers.git/tree/net-print/brother-mfc6490cw-cups/metadata.xml)).

# Tasks to confirm/verify

- [X] **Print a pdf with lpr** from a PC (via cups daemon... but with little ink in the cartridges :sweat_smile:).
- [X] **Run /usr/bin/brprintconf_mfc6490cw** to see if the path /opt/brother/Printers/... is used (see _Patching binary files_.
- [ ] Should the **avahi** use flag be implemented as it is for brother-mfc9130cw-bin?
- [ ] Is it ok to leave both **amd64** and **x86** as keywords, or just amd64 shall be used?

# Details

This ebuild merges [net-print/brother-mfc6490cw-cups](https://cgit.gentoo.org/proj/printer-drivers.git/tree/net-print/brother-mfc6490cw-cups) and [net-print/brother-mfc6490cw-lpr](https://cgit.gentoo.org/proj/printer-drivers.git/tree/net-print/brother-mfc6490cw-lpr) drivers:
- Update **Copyright**'s date
- Update **EAPI**
- Update **MY_PN**
- Update **DESCRIPTION**
- Update **HOMEPAGE**
- Provide both CUPS and LPR **SRC_URI**s
- Provide both CUPS and LPR **LICENSE**s (_changed LPR's **Brother** license to **brother-eula**_)
- Combine CUPS and LPR **RDEPENDS**
- Block **net-print/brother-mfc6490cw-cups** and **net-print/brother-mfc6490cw-lpr**
- Update **FILTER_PATH** to point directly to **/usr/libexec/cups/filter** in the workdir
- Install both CUPS and LPR into **/opt/brother/Printers** (also _patch binaries **brprintconf_mfc6490cw** and **brmfc6490cwfilter**_, see the subsection _Patching binary files_)
- Install brprintconf_mfc6490cw into **/usr/bin** (as usual)
- Still uses the **/usr/share/cups/model** directory

## Patching binary files

`/usr/local/Brother/Printer` was changed into `/opt/brother/Printers`, which is the default directory used by many other net-printer's Brother drivers, like brother-mfc9130cw-bin.

Here is illustrated the method used to modify the binary files (_the full path and name of files may be omitted for brevity_):
```
$ objcopy --dump-section .rodata=brprintconf_mfc6490cw.dump brprintconf_mfc6490cw
$ objcopy --dump-section .rodata=brmfc6490cwfilter.dump brmfc6490cwfilter
$ dhex brprintconf_mfc6490cw.dump (search/replace the location as specified before)
$ dhex brmfc6490cwfilter.dump (search/replace the location as specified before)
$ cp brprintconf_mfc6490cw.dump brmfc6490cwfilter.dump net-print/brother-mfc6490cw-bin/files
```
The dumps are restored with something like the following (**see the ebuild for exact commands**):
```
$ objcopy --update-section .rodata=brprintconf_mfc6490cw.dump brprintconf_mfc6490cw
```

### CODE REVIEW

**Comment by mehw** [comment]
Hi @stefan-langenmaier,

I see there are a lot of Travis failures due to some drivers in this overlay.

If I'm not mistaken, **brother-mfc6490cw-bin** fails due to a **file.size-fatal**:
```
(92 KiB) net-print/brother-mfc6490cw-bin/files/brmfc6490cwfilter-1.1.2_p2.rodata
```
I have no error emerging the ebuild on my PC. I don't get it, I'm not aware of any file size limit.

What do you suggest?

**Comment by stefan-langenmaier** [comment]
Thanks for the explanation concerning the binary patching. I think I can see now the issue with the binary file. If I understand correctly the `dhex` step is difficult to automate.

File: net-print/brother-mfc6490cw-bin/brother-mfc6490cw-bin-1.1.2_p2.ebuild [hunk 1]
```diff
@@ -0,0 +1,103 @@
+# Copyright 1999-2019 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+# $Header: $
```
[the remark this answers is not in the document]
Comment (mehw) [reply net-print/brother-mfc6490cw-bin/brother-mfc6490cw-bin-1.1.2_p2.ebuild]: Thanks! I addressed that in a new commit.

**Comment by mehw** [comment]
> First of all thanks for this detailed pull request.

You're welcome!

> My only real question is why are you adding the binary files, when the files could also be generated?

LPR's `brmfc6490cwfilter` and `brprintconf_mfc6490cw` are closed source. They have a hardcoded path pointing to `/usr/local/Brother/Printer`. To be consistent with other drivers in net-printer, I've changed that to `/opt/brother/Printers`.

> If I understand correctly the `dhex` step is difficult to automate.

I did use `dhex` only to create the patches for the binary files from above. The hardcoded path is in the .rodata section, `objcopy` can dump that easily, then I did a **manual edit** to the dumped sections, and the patching of the LPR binary files is done by `objcopy` from the ebuild importing back the edited dumped sections which reside in the `files` directory.

**Comment by mehw** [comment]
**`cupswrapper`** is now built from GPLv2 source code.

 Also, the new way to [patch LPR binary files](https://github.com/stefan-langenmaier/brother-overlay/pull/75#issuecomment-546148321), done directly form the ebuild itself via `printf` and `objcopy`, should solve the [Travis file.size-fatal failure](https://github.com/stefan-langenmaier/brother-overlay/pull/75#issuecomment-545520158) error.

I did a couple of forced pushes to fix some Travis complaints. First a `digest.unused` and then a `ebuild.minorsyn` (unquoted variable).

I see that there are a lot of `ebuild.badheader` errors like the following:
```
media-gfx/brother-ads1000w-bin/brother-ads1000w-bin-1.0.ebuild: Stale CVS header on line: 3
```
What do you think @stefan-langenmaier? IMHO those should be easily fixable, as I did before [following your advice](https://github.com/stefan-langenmaier/brother-overlay/pull/75#discussion_r338223318).

**Comment by stefan-langenmaier** [comment]
First of all, this is amazing work! I'm learning a lot.
Thanks for putting in the effort to avoid the binary files and patching the files directly in the ebuild. This will be a lot clearer to the next user.

Also *very* cool to use the source for the cups driver.
I hope that one day some of the drivers can be used on other architectures/we can avoid the binary files. But you are right
>  IMHO those should be easily fixable [...]

there are other things to fix at the moment with the repository. These issues should be easily fixable. Unfortunately I didn't tend enough to the repo in recent times. If you want to look into cleaning this up a little bit, I'd be very happy, but again thanks a lot for this very thorough PR.

I think an even more important step will be to write some documentation/structure/eclasses to share common functionality between ebuilds as they have drifted a part/there are by now a lot more than I have anticipated. :) Your ebuild can be a good source/template for this.

**Comment by mehw** [comment]
Thanks for merging!

> I hope that one day some of the drivers can be used on other architectures/we can avoid the binary files.

You are right, the drivers should be compiled from source _whenever_ it's possible. But unfortunately, the source code is not always made available...

> If you want to look into cleaning this up a little bit, I'd be very happy, but again thanks a lot for this very thorough PR.

I may look into it, at least the more trivial stuff.

> I think an even more important step will be to write some documentation/structure/eclasses to share common functionality between ebuilds as they have drifted a part/there are by now a lot more than I have anticipated. :)

That will be cool! I have a lot to learn here. I hope you are not in a hurry. This is an incentive to do some experiment. Who knows what may sprout...

Fixing Travis errors may be a start.

Cheers!

**Pull request merged**